### PR TITLE
Use Build Kit docker feature to shrink the Docker image size

### LINKED ISSUE #3666: The distribution Dockerfiles are larger than they need to be
**Is your feature request related to a problem? Please describe.**
The distribution Docker builds add install artifacts for Heron and Zookeeper. The files are deleted in later layers, but the size of the resulting image is still large. The user can use `--squash` to compress the layers and reclaim the size, but if we used multi-stage Docker builds, we could automatically make the images smaller.

**Describe the solution you'd like**
Use Docker multi-stage builds to install components in a base stage, and then copy over the resulting artifacts into the later stage.

**Describe alternatives you've considered**
Docker `--squash` which helps if the Docker user has the experimental feature enabled. This loses layer information.

### PR DESCRIPTION
**Pull request by nicknezis**
Resolves #3666 

The `heron-install.sh` file was being left in a Docker layer after the install. This is causing the Heron distribution image to be about 2.3 GB in size. This change helps drop the dize to roughly 1.1 GB.

### CODE REVIEW

**Review by Code0x58 — APPROVED** [review]
Requiring [`docker buildx`](https://docs.docker.com/buildx/working-with-buildx/#install) for builds sounds reasonable to me as people may usually have it from their reasonably up to date docker installation.

I tried to test the build process with the command below but it highlighted that debian9 doesn't have the required version of python, but the others built normally.
```shell
./scripts/release/docker-images "$(git describe --tags)" --all
```

It looks like there's ~190M of additional space which could be cut out of the heron install due to file duplication, I'll make an issue for that in case anyone feels it's worth picking up.

File: docker/dist/Dockerfile.dist.centos7 [hunk 1]
```diff
@@ -34,14 +34,15 @@ RUN yum -y install epel-release \
 
 ENV JAVA_HOME /usr/
 
-ADD artifacts /heron
+# run Heron installer
+RUN --mount=type=bind,source=artifacts,target=/tmp/heron /tmp/heron/heron-install.sh && \
+    rm /usr/local/heron/dist/heron-core.tar.gz && \
+    rm -rf /opt/heron/heron-core/lib/scheduler/heron-local-scheduler.jar && \
+    rm -rf /opt/heron/heron-core/lib/scheduler/heron-mesos-scheduler.jar && \
+    rm -rf /opt/heron/heron-core/lib/scheduler/heron-slurm-scheduler.jar
```
[the remark this answers is not in the document]
Comment (nicknezis) [reply docker/dist/Dockerfile.dist.centos7:42]: Yes I agree. I saw the lines in one of the Dockerfiles and made them consistent. But I don't know if I like this. It assumes the container if only used for Kubernetes. I'm not sure if that will always be the case. I can remove these `rm` statements.

[the remark this answers is not in the document]
Comment (nicknezis) [reply docker/dist/Dockerfile.dist.centos7:42]: Would you agree that removing `heron-core.tar.gz` might still be worthwhile?

[the remark this answers is not in the document]
Comment (Code0x58) [reply docker/dist/Dockerfile.dist.centos7:42]: I imagine as a `.tar.gz` it isn't of use at runtime, so probably fair to strip. I wouldn't be surprised if it's just an artefact that was packed into `heron-install.sh` and it was just unpacked but not cleaned up.

[the remark this answers is not in the document]
Comment (nicknezis) [reply docker/dist/Dockerfile.dist.centos7]: It seems that some schedulers can be configured to point to a hosted `heron-core.tar.gz` and the individual nodes will download the file to provide the Heron framework files locally. But in a framework like Kubernetes where the jars are already available in the container, there is no need for the extra tar.gz. I wonder if it's possible to add other config flags on `heron-install.sh` to toggle what actually gets installed.

For now I think what we have is a good compromise. No `tar.gz`, but I'll keep the other scheduler jars.

**Review by joshfischer1108 — APPROVED** [review]
Do we need to make any changes to the documentation to account for `buildx`?

**Comment by nicknezis** [comment]
> Do we need to make any changes to the documentation to account for `buildx`?

That's a good question. I did a recursive grep through the repo for ``

**Pull request merged**